docs/dev_docs.md

```diff
@@ -1,3 +1,449 @@
+        Reset needs to initialize the following attributes
+        - agents
+        - rewards
+        - _cumulative_rewards
+        - dones
+        - infos
+        - agent_selection
+        And must set up the environment so that render(), step(), and observe()
+        can be called without issues.
+
+        Here it sets up the state dictionary
+        which is used by step() and the observations dictionary which is used by step()
+        and observe()
+        '''
+        self.agents = self.possible_agents[:]
+        self.rewards = {agent: 0 for agent in self.agents}
+        self._cumulative_rewards = {agent: 0 for agent in self.agents}
+        self.dones = {agent: False for agent in self.agents}
+        self.infos = {agent: {} for agent in self.agents}
+        self.state = {agent: NONE for agent in self.agents}
+        self.observations = {agent: NONE for agent in self.agents}
+        self.num_moves = 0
+        '''
+        Our agent_selector utility allows easy cyclic stepping through the agents list.
+        '''
+        self._agent_selector = agent_selector(self.agents)
+        self.agent_selection = self._agent_selector.next()
+
+    def step(self, action):
+        '''
+        step(action) takes in an action for the current agent (specified by
+        agent_selection) and needs to update
+        - rewards
+        - _cumulative_rewards (accumulating the rewards)
+        - dones
+        - infos
+        - agent_selection (to the next agent)
+        And any internal state used by observe() or render()
+        '''
+        if self.dones[self.agent_selection]:
+            # handles stepping an agent which is done
+            # takes in the None action, and moved agent_selection to the next
+            # done agent or if there are no more done agents, to the next live agent
+            return self._was_done_step(action)
+
+        agent = self.agent_selection
+
+        # the agent which stepped last had its _cumulative_rewards accounted for
+        # (because it was returned by last()), so the _cumulative_rewards for this
+        # agent should start again at 0
+        self._cumulative_rewards[agent] = 0
+
+        # stores action of current agent
+        self.state[self.agent_selection] = action
+
+        # collect reward if it is the last agent to act
+        if self._agent_selector.is_last():
+            # rewards for all agents are placed in the .rewards dictionary
+            self.rewards[self.agents[0]], self.rewards[self.agents[1]] = {
+                (ROCK, ROCK): (0, 0),
+                (ROCK, PAPER): (-1, 1),
+                (ROCK, SCISSORS): (1, -1),
+                (PAPER, ROCK): (1, -1),
+                (PAPER, PAPER): (0, 0),
+                (PAPER, SCISSORS): (-1, 1),
+                (SCISSORS, ROCK): (-1, 1),
+                (SCISSORS, PAPER): (1, -1),
+                (SCISSORS, SCISSORS): (0, 0),
+            }[(self.state[self.agents[0]], self.state[self.agents[1]])]
```

Comment:
Purely for the code quality of RPS, this dict could be defined at the top of the file